dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpCdiExtension.java

```diff
@@ -23,20 +28,44 @@
 public class McpCdiExtension implements Extension {
 
     private ToolRegistry tools = new ToolRegistry();
+    private ConcurrentHashMap<String, LinkedList<String>> duplicateToolsMap = new ConcurrentHashMap<>();
 
     void registerTools(@Observes ProcessManagedBean<?> pmb) {
         // TODO: limit this to just bean types with Tool annotations?
         AnnotatedType<?> type = pmb.getAnnotatedBeanClass();
+        Class<?> javaClass = type.getJavaClass();
         for (AnnotatedMethod<?> m : type.getMethods()) {
             Tool toolAnnotation = m.getAnnotation(Tool.class);
             if (toolAnnotation != null) {
-                registerTool(toolAnnotation, pmb.getBean(), m);
+                registerTool(toolAnnotation, pmb.getBean(), m, javaClass.getPackageName() + "." + javaClass.getSimpleName());
             }
         }
     }
 
-    private void registerTool(Tool tool, Bean<?> bean, AnnotatedMethod<?> method) {
-        tools.addTool(ToolMetadata.createFrom(tool, bean, method));
+    void afterDeploymentValidation(@Observes AfterDeploymentValidation afterDeploymentValidation, BeanManager manager) {
+
+        // prune items that are not duplicates
+        duplicateToolsMap.entrySet().removeIf(e -> e.getValue().size() == 1);
+        StringBuilder sb = new StringBuilder();
+
+        for (String toolName : duplicateToolsMap.keySet()) {
+            LinkedList<String> qualifiedNames = duplicateToolsMap.get(toolName);
+            sb.append("Tool: ").append(toolName);
+            sb.append(" -- Methods found:\n");
+            for (String qualifiedName : qualifiedNames) {
+                sb.append("    - ").append(qualifiedName + "\n");
+            }
+        }
+
+        if (duplicateToolsMap.size() != 0) {
+            afterDeploymentValidation.addDeploymentProblem(new Exception("More than one MCP tool has the same name: \n" + sb.toString()));
+        }
+    }
```

Comment:
Try to avoid concatenation when you're already using a StringBuilder

```suggestion
    void afterDeploymentValidation(@Observes AfterDeploymentValidation afterDeploymentValidation, BeanManager manager) {

        // prune items that are not duplicates
        duplicateToolsMap.entrySet().removeIf(e -> e.getValue().size() == 1);

        if (duplicateToolsMap.size() != 0) {
            StringBuilder sb = new StringBuilder("More than one MCP tool has the same name: \n");

            for (String toolName : duplicateToolsMap.keySet()) {
                LinkedList<String> qualifiedNames = duplicateToolsMap.get(toolName);
                sb.append("Tool: ").append(toolName);
                sb.append(" -- Methods found:\n");
                for (String qualifiedName : qualifiedNames) {
                    sb.append("    - ").append(qualifiedName).append("\n");
                }
            }

            afterDeploymentValidation.addDeploymentProblem(new Exception(sb.toString()));
        }
    }
```